[stable/stolon] add stolon chart

rtluckie opened the pull request:
Initial commit of stolon chart.

Commit:
[stable/stolon] add stolon chart

rtluckie commented:
/assign @unguiculus

rtluckie commented:
/assign @prydonius

unguiculus commented:
/assign

timstoop commented on stable/stolon/values.yaml:
```diff
@@ -0,0 +1,86 @@
+image:
+  repository: sorintlab/stolon
+  tag: v0.10.0-pg9.6
+  pullPolicy: Always
```
I believe we prefer to stick to `IfNotPresent` by default.

rtluckie replied:
fixed.

timstoop commented on stable/stolon/values.yaml:
```diff
@@ -0,0 +1,86 @@
+# superuserPassword: ""
+
+store:
+  backend: "consul"
```
This differs from the value set in the README. I'm also not sure if either consul or etcd will get this through CI, as the Chart does not deploy them nor does it set CI overrides to install them. This might be an issue once this is ok-to-test.

rtluckie replied:
Updated README and included consul as a requirement.

timstoop commented on stable/stolon/values.yaml:
```diff
@@ -0,0 +1,86 @@
+  replicaCount: 2
+  resources:
+    requests:
+      cpu: "100m"
```
We prefer charts not to set defaults for resources.

rtluckie replied:
fixed.

timstoop reviewed (changes requested):
Nice to see how complete your submission is regarding default features!

unguiculus commented on stable/stolon/templates/_helpers.tpl:
```diff
@@ -0,0 +1,73 @@
+{{- printf $name -}}
+{{- end -}}
+
+{{- define "stolon.keeper.fullname" -}}
```
You defined partials that are not used.

rtluckie replied:
fixed.

unguiculus commented on stable/stolon/templates/create-cluster-job.yaml:
```diff
@@ -0,0 +1,50 @@
+              {
+                  "name": "create-cluster",
+                  "image": "gcr.io/google_containers/etcd-amd64:3.0.14",
+                  "command": ["/bin/sh", "-c", "while ! etcdctl --endpoints {{ .Values.store.endpoints }} cluster-health; do sleep 1 && echo -n .; done"],
```
I quickly look at the docs and think you can run Stolon without Etcd on Kubernetes 1.8+. In any case you should clearly document the requirements.

rtluckie replied:
fixed.  using kubernetes as default store.

unguiculus commented on stable/stolon/templates/create-cluster-job.yaml:
```diff
@@ -0,0 +1,50 @@
+                  "name": "create-cluster",
+                  "image": "gcr.io/google_containers/etcd-amd64:3.0.14",
+                  "command": ["/bin/sh", "-c", "while ! etcdctl --endpoints {{ .Values.store.endpoints }} cluster-health; do sleep 1 && echo -n .; done"],
+                  "imagePullPolicy": {{ default "Always" .Values.imagePullPolicy | quote }}
```
Put defaults in `values.yaml` instead of using the `default` function.

rtluckie replied:
fixed.

unguiculus commented on stable/stolon/templates/create-secrets-job.yaml:
```diff
@@ -0,0 +1,34 @@
+          command:
+          - sh
+          - -c
+          - "if ! kubectl get secret generic {{ template "stolon.fullname" . }}; then kubectl create secret generic {{ template "stolon.fullname" . }} --namespace {{ .Release.Namespace }} --from-literal=pg_su_password={{- if .Values.superuserPassword -}}'{{ .Values.superuserPassword }}' {{ else }}'{{ randAlphaNum 40 }}' {{ end }} --from-literal=pg_repl_password={{- if .Values.replicationPassword  ...
```
Why do you use kubectl to create a secret insteaf of managing this with the chart?

rtluckie replied:
I initially attempted to manage the secrets in the chart, but found that the secrets (specifically from `randAlphaNum`) change on `helm upgrade`.

This is a hack to ensure that secrets persist after`helm upgrade`

unguiculus commented on stable/stolon/templates/keeper-statefulset.yaml:
```diff
@@ -0,0 +1,96 @@
+apiVersion: apps/v1beta1
```
You should go with `apps/v1beta2`

rtluckie replied:
fixed.

unguiculus commented on stable/stolon/templates/keeper-statefulset.yaml:
```diff
@@ -0,0 +1,96 @@
+        app: {{ template "stolon.name" . }}-keeper
+        release: {{ .Release.Name }}
+      annotations:
+        pod.alpha.kubernetes.io/initialized: "true"
```
This annotation is obsolete as of Kubernetes 1.8. Remove it.

rtluckie replied:
removed.

unguiculus commented on stable/stolon/templates/keeper-statefulset.yaml:
```diff
@@ -0,0 +1,96 @@
+          secret:
+            secretName: {{ template "stolon.fullname" . }}
+
+{{- if .Values.persistence.enabled }}
```
Make sure this confirms with our best practices. See review guidelines.

rtluckie replied:
fixed.

unguiculus commented on stable/stolon/templates/proxy-deployment.yaml:
```diff
@@ -0,0 +1,68 @@
+apiVersion: extensions/v1beta1
```
apps/v1beta2

rtluckie replied:
fixed.

unguiculus commented on stable/stolon/templates/keeper-statefulset.yaml:
```diff
@@ -0,0 +1,96 @@
+    chart: {{ template "stolon.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
```
You need to add the service name

rtluckie replied:
fixed.

unguiculus commented on stable/stolon/templates/keeper-service.yaml:
```diff
@@ -0,0 +1,21 @@
+{{ toYaml . | indent 4 }}
+{{- end }}
+spec:
+  type: {{ .Values.keeper.service.type }}
```
Statefulsets require a headless governing service.

rtluckie replied:
Fixed.

unguiculus commented on stable/stolon/templates/sentinel-deployment.yaml:
```diff
@@ -0,0 +1,59 @@
+apiVersion: extensions/v1beta1
```
apps/v1beta2

rtluckie replied:
fixed.

unguiculus commented on stable/stolon/values.yaml:
```diff
@@ -0,0 +1,86 @@
+  # create certificates according to these instructions: https://www.postgresql.org/docs/9.6/static/ssl-tcp.html
+  # to enable encrypted traffic, servert.crt and server.key are required, by that name.
+  # the use of ** Client Certificates ** is not supported
+  client_ssl:
```
Camel case only, please.

rtluckie replied:
fixed.

Commit:
Requested changes from reviewers

- update image.pullPolicy
- update backend to use consul by default
- remove resource specs
- remove unused partials
- update apiVersion
- remove obsolete annotation
- update persistence spec to align with best practices
- fix case in values.yaml

Commit:
don't install etcd by default

Commit:
Refactor service and port definitions

- refactor service manifests to allow more granular configuration
- update port definitions in values.yaml

Commit:
Numerous fixes for issues found during testing

- remove stolon.clusterName partial from _helpers.tpl and change manifests accordingly
- fix create-secrets-job
- refactor labels and selectors in all manifests to align with best practices
- add keeper-headless-service for statefulset

timstoop commented:
/ok-to-test

timstoop commented:
As @unguiculus says, it looks like Stolon can work directly with Kubernetes. If you change the values.yaml to use that by default instead, I think it should pass the CI. Otherwise you'll have to provide additional manifests for the CI to provide it with the additional services. Check this document for additional guidance for the latter: https://github.com/kubernetes/charts/blob/master/test/README.md

Commit:
Changes to fix kubernetes as backend store

- update label.app to support stolon requirements
  - see https://github.com/sorintlab/stolon/commit/38ae6b13b5e161a5bfe0fbe01084ca060eaf2e76#diff-95cdd374e9440fde010ff35d65f8cf3fR54
- add .Values.store.kubeResourceKind
- make kubernetes default store in Values.yaml
- remove subcharts for consul and etcd
- add nodeSelector, affinity and toleration for keeper-statefulset
- clean up values.yaml
- update README to reflect changes

rtluckie commented:
@unguiculus @timstoop All issues/requested changes should be fixed.

unguiculus commented on stable/stolon/templates/keeper-service.yaml:
```diff
@@ -0,0 +1,36 @@
+  {{end}}
+{{- else if eq .Values.keeper.service.type "LoadBalancer" }}
+  type: {{ .Values.keeper.service.type }}
+  loadBalancerIP: {{ default "" .Values.keeper.service.loadBalancerIP }}
```
Remove `default` function call. Defaults should go into `values.yaml`.

rtluckie replied:
fixed

unguiculus commented on stable/stolon/templates/keeper-service.yaml:
```diff
@@ -0,0 +1,36 @@
+{{ toYaml . | indent 4 }}
+{{- end }}
+spec:
+{{- if (or (eq .Values.keeper.service.type "ClusterIP" "") (empty .Values.keeper.service.type)) }}
```
This should probably be like so:
`{{- if (or (eq .Values.keeper.service.type "ClusterIP") (empty .Values.keeper.service.type)) }}`

rtluckie replied:
fixed

unguiculus commented on stable/stolon/templates/_helpers.tpl:
```diff
@@ -0,0 +1,49 @@
+Create a default fully qualified store name.
+We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
+*/}}
+{{- define "stolon.store.fullname" -}}
```
This is still there but not used. Please remove it.

rtluckie replied:
fixed

unguiculus commented on stable/stolon/templates/create-secrets-job.yaml:
```diff
@@ -0,0 +1,34 @@
+          command:
+          - sh
+          - -c
+          - "if ! kubectl get secret {{ template "stolon.fullname" . }}; then kubectl create secret generic {{ template "stolon.fullname" . }} --namespace {{ .Release.Namespace }} --from-literal=pg_su_password={{- if .Values.superuserPassword -}}'{{ .Values.superuserPassword }}' {{ else }}'{{ randAlphaNum 40 }}' {{ end }} --from-literal=pg_repl_password={{- if .Values.replicationPassword -}}'{{ . ...
```
Please don't use this hack to crate the secret. Many charts use auto-generated password and have issues with update. Real-world installations should not use auto-generated passwords. You can document this in the readme if you like. Alternatively, you could use `.Release.IsInstall` to conditionally install the secret only when the chart is first installed but not during an update.

rtluckie replied:
fixed

unguiculus commented on stable/stolon/templates/keeper-statefulset.yaml:
```diff
@@ -0,0 +1,117 @@
+    heritage: {{ .Release.Service }}
+    component: keeper
+spec:
+  serviceName: {{ template "stolon.fullname" . }}-keeper
```
`{{ template "stolon.fullname" . }}-keeper-headless`

rtluckie replied:
fixed

unguiculus commented on stable/stolon/templates/keeper-statefulset.yaml:
```diff
@@ -0,0 +1,117 @@
+metadata:
+  name: {{ template "stolon.fullname" . }}-keeper
+  labels:
+    app: stolon-keeper
```
The `app` label should be `{{ template "stolon.name" . }}` throughout the chart.

rtluckie replied:
fixed.

unguiculus commented on stable/stolon/templates/keeper-statefulset.yaml:
```diff
@@ -0,0 +1,117 @@
+      app: stolon-keeper
+      release: {{ .Release.Name }}
+      component: keeper
+      stolon-cluster: {{ template "stolon.fullname" . }}
```
What is label `stolon-cluster` needed for? Anyways, it is not required to be part of selectors.

rtluckie replied:
removed `stolon-cluster` from selectors.

stolon uses `stolon-cluster`label for component discovery [link](https://github.com/sorintlab/stolon/blob/master/doc/architecture.md#kubernetes-store-backend)

unguiculus commented on stable/stolon/templates/proxy-service.yaml:
```diff
@@ -0,0 +1,36 @@
+{{ toYaml . | indent 4 }}
+{{- end }}
+spec:
+{{- if (or (eq .Values.proxy.service.type "ClusterIP" "") (empty .Values.proxy.service.type)) }}
```
This should probably be like so:
`{{- if (or (eq .Values.proxy.service.type "ClusterIP") (empty .Values.proxy.service.type)) }}`

rtluckie replied:
fixed

Commit:
Requsted changes, cleanup, misc fixes

- update image to latest version of stolon
- remove create-secrets-job and use native secrets w/ {{ if .Release.IsInstall }} to avoid overwriting secrets
- remove unnecessary keeper service
- remove unused template
- cleanup labels and selectors
- add component and stolon-cluster labels to support discovery in latest version of stolon
- add annotations to services, deployment and statefulset
- enable stolon metrics

Commit:
Add pgParameters option

- add ability to add postgresql.conf config options
- update create-cluster-job to use new config options
- remove deprecated config options
- update README and values.yaml
- move to master-pg10 image tag

timstoop commented on stable/stolon/README.md:
```diff
@@ -0,0 +1,64 @@
+| Parameter                               | Description                                    | Default                                                      |
+| --------------------------------------- | ---------------------------------------------- | ------------------------------------------------------------ |
+| `image.repository`                      | `stolon` image repository                      | `sorintlab/stolon`                                           |
+| `image.tag`                             | `stolon` image tag                             | `master-pg10`                                               |
```
You changed this from 0.10. I'd prefer to have a versioned image instead of a "rolling-update" one. Any reason why you switched?

rtluckie replied:
I needed the latest release to fix some issues. previous version of stolon (0.10) had some very opinionated requirements re: app name, label, etc. We should move to a specific image tag as soon as it is available.

ATM, the master image is 21 days old [link](https://hub.docker.com/r/sorintlab/stolon/tags/)

unguiculus replied:
We require stable version. Any idea when the next release is due?

rtluckie replied:
@unguiculus @timstoop  I was incorrect about needing to use the master image version.  Reverted to stable version and everything looks to be working.

unguiculus commented on stable/stolon/templates/create-cluster-job.yaml:
```diff
@@ -0,0 +1,50 @@
+      restartPolicy: OnFailure
+      serviceAccountName: {{ template "stolon.serviceAccountName" . }}
+      containers:
+      - name: {{ template "stolon.fullname" . }}
```
Please make list indentation consistent throughout the chart. Since you mostly indent two spaces, please do so as well here.

rtluckie replied:
fixed

unguiculus commented on stable/stolon/templates/create-cluster-job.yaml:
```diff
@@ -0,0 +1,50 @@
+        release: {{ .Release.Name }}
+      annotations:
+        {{- if eq .Values.store.backend "etcd" }}
+        pod.beta.kubernetes.io/init-containers: '[
```
Don't use annotations for init container any more.

rtluckie replied:
fixed

unguiculus commented on stable/stolon/templates/keeper-statefulset.yaml:
```diff
@@ -0,0 +1,120 @@
+  selector:
+    matchLabels:
+      app: {{ template "stolon.name" . }}
+      release: {{ .Release.Name }}
```
The selector needs the component label.

rtluckie replied:
fixed

unguiculus commented on stable/stolon/templates/proxy-deployment.yaml:
```diff
@@ -0,0 +1,82 @@
+  selector:
+    matchLabels:
+      app: {{ template "stolon.name" . }}
+      release: {{ .Release.Name }}
```
The selector needs the component label.

rtluckie replied:
fixed

unguiculus commented on stable/stolon/templates/sentinel-deployment.yaml:
```diff
@@ -0,0 +1,71 @@
+  selector:
+    matchLabels:
+      app: {{ template "stolon.name" . }}
+      release: {{ .Release.Name }}
```
The selector needs the component label.

rtluckie replied:
fixed

Commit:
Requested changes

- revert to use stable stolon image
- add etcdImage to values
- fix indent
- remove initContainer annotation and replace with proper spec
- add component label to selector specs

Commit:
Fix resource definitions

rtluckie commented:
@unguiculus @timstoop let me know if other changes are needed.

unguiculus commented:
/lgtm
